File: PART3/ch02/ch02_01/java-kotlin/src/main/java/org/example/ex08/Exam08.java
package org.example.ex08;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Optional;

public class Exam08 {

    public Exam08(Store store){    // service logic
        var stringRegisterAt = toLocalDateTimeString(
            Optional.ofNullable(store.getRegisterAt())
        );
    }

    public String toLocalDateTimeString(Optional<LocalDateTime> localDateTime){

        LocalDateTime temp = localDateTime.orElseGet(()-> LocalDateTime.now());
        return temp.format(DateTimeFormatter.ofPattern("yyyy MM dd"));
    }

    public static void main(String[] args){
        // client ->
        var registerDto = new Store();
        //registerDto.setRegisterAt(LocalDateTime.now());

        new Exam08(registerDto);    // example service logic
    }
}

class Store{
    private LocalDateTime registerAt;

    public LocalDateTime getRegisterAt() {
        return registerAt;
    }

    public void setRegisterAt(LocalDateTime registerAt) {
        this.registerAt = registerAt;
    }
}

File: PART3/ch01/kotlin-example/kotlin-example/src/main/kotlin/ex08/Exam08.kt
package ex08

import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

fun main(){

    Exam08(Store())

    var now = LocalDateTime.now()
    println(DateTimeUtil().localDateTimeToString(now))
}

data class Store(
    var registerAt: LocalDateTime ?= null
)

class Exam08{

    constructor(store: Store){
        val strLocalDateTime = toLocalDateTimeString(store.registerAt)
        println(strLocalDateTime)

        println(toLocalDateTimeString())
    }

    fun toLocalDateTimeString(
        localDateTime: LocalDateTime? = LocalDateTime.of(2020,2,2,13,0,0)
    ): String{
        val temp = localDateTime?: LocalDateTime.now()
        return temp.format(DateTimeFormatter.ofPattern("yyyy MM dd"))
    }
}

class DateTimeUtil {

    private val KST_FORMAT = "yyyy년 MM월 dd일 HH시 mm분 ss초"

    fun localDateTimeToString(
        localDateTime: LocalDateTime? = LocalDateTime.now(),
        pattern: String ?= KST_FORMAT
    ): String {
        return (localDateTime?: LocalDateTime.now()).format(DateTimeFormatter.ofPattern(pattern))
    }

}